src/main/java/org/gradlex/javamodule/moduleinfo/ModuleDescriptorRecommendation.java

```diff
@@ -0,0 +1,326 @@
+
+        final Set<ModuleIdentifier> compileDependencies = new HashSet<>();
+
+        final File jar;
+
+        final SortedSet<String> requires = new TreeSet<>();
+        final SortedSet<String> requiresTransitive = new TreeSet<>();
+        final SortedSet<String> requiresStatic = new TreeSet<>();
+        final SortedSet<String> exports = new TreeSet<>();
+
+        final SortedSet<String> provides = new TreeSet<>();
+
+        String moduleName;
+
+        boolean automatic;
+
+        Artifact(ModuleIdentifier coordinates, File jar) {
+            this.coordinates = coordinates;
+            this.jar = jar;
+        }
+
+        Set<ModuleIdentifier> allDependencies() {
+            Set<ModuleIdentifier> out = new HashSet<>();
+            out.addAll(compileDependencies);
+            out.addAll(runtimeDependencies);
+            return out;
+        }
+
+        boolean containsAnyRequires(String moduleName) {
+            return requires.contains(moduleName) || requiresTransitive.contains(moduleName) || requiresStatic.contains(moduleName);
+        }
+
+        String dsl() {
+            List<String> out = new ArrayList<>();
+            String group = this.coordinates.getGroup();
+            String name = this.coordinates.getName();
+            String moduleName = this.moduleName;
+            out.add("module('" + group + ":" + name + "', '" + moduleName + "') {");
+            out.add("    closeModule()");
+            for (String item : this.requiresTransitive) {
+                out.add("    requiresTransitive('" + item + "')");
+            }
+            for (String item : this.requiresStatic) {
+                out.add("    requiresStatic('" + item + "')");
+            }
+            for (String item : this.requires) {
+                out.add("    requires('" + item + "')");
+            }
+            for (String item : this.exports) {
+                out.add("    exports('" + item + "')");
+            }
+            for (String item : this.provides) {
+                out.add("    // ignoreServiceProvider('" + item + "')");
+            }
+            out.add("}");
+            return String.join("\n", out)
+                    .replace('\'', '"');
+        }
+
+    }
+
+    interface Java8SafeToolProvider {
+
+        int run(PrintWriter out, PrintWriter err, String... args);
+
+        @SuppressWarnings("Since15")
+        static Java8SafeToolProvider findFirst(String name) {
+            try {
+                ToolProvider tool = ToolProvider.findFirst(name)
+                        .orElseThrow(() -> new RuntimeException("The JDK does not bundle " + name));
+                return tool::run;
+            } catch (NoClassDefFoundError e) {
+                throw new RuntimeException("This functionality requires Gradle to be powered by JDK 11+", e);
+            }
+        }
+
+    }
+
+    @Input
+    public abstract Property<Configuration> getRuntimeConfiguration();
```

Comment:
This is not compatible with Gradle's configuration cache. The tests currently do not cover this by default, but you see the problem if you add the `--configuration-cache` flag to the runner in the fixture and then run the tests in `RecommendModuleSpecFunctionalTest`.

This can be solved by passing the _Files_ and metadata as separate inputs to the task. Because _files_ and _data_ is handled differently in task inputs. This unfortunately makes it a bit more complicated. I have done a working solution here, because it is hard to explain in words:
https://github.com/gradlex-org/extra-java-module-info/commit/ee89e093d808210f1ba35a6f6167cb38cd254ed1

Maybe you can take that and clean in up a bit.

The important bit is to have two lists that go together and follow a certain order. This is not really nice, but I am not aware of a better solution at the moment. This is based on this example from Gradle: https://github.com/adammurdoch/dependency-graph-as-task-inputs

Reply:
This is a great catch! I slightly polished your implementation and incorporated it into this PR.